src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSyncController.scala

```diff
@@ -0,0 +1,181 @@
+  import Config.FastSync._
+
+  private implicit val sender: ActorRef = context.self
+
+  def regularSync(log: LoggingAdapter): Actor.Receive = {
+
+    case StartRegularSync =>
+      askForHeaders(log)
+
+    case ResumeRegularSync =>
+      askForHeaders(log)
+
+    case MessageReceived(m: BlockHeaders) if !resolvingBranch =>
+      handleDownload(m, log)
+
+    case MessageReceived(m: BlockHeaders) if resolvingBranch =>
+      handleBlockBranchResolution(m, log)
+
+    case MessageReceived(m: BlockBodies) =>
+      handleBlockBodies(m, log)
+
+    case m: BroadcastBlocks if broadcasting =>
+      peersToDownloadFrom.keys.foreach(_ ! m)
+
+    case ResolveBranch(peer) =>
+      resolvingBranch = true
+      peer ! SendMessage(GetBlockHeaders(Right(headers.head.parentHash), blockResolveDepth, skip = 0, reverse = true))
+  }
+
+  private def askForHeaders(log: LoggingAdapter) = {
+
+    bestPeer match {
+      case Some(peer) =>
+        val blockNumber = appStateStorage.getBestBlockNumber()
+
+        peer ! Subscribe(Set(BlockHeaders.code, BlockBodies.code))
+        peer ! SendMessage(GetBlockHeaders(Left(blockNumber + 1), blockHeadersPerRequest, skip = 0, reverse = false))
+      case None =>
+        log.warning("no peers to download from")
+        scheduleResume
+    }
+  }
+
+  private def handleBlockBranchResolution(message: BlockHeaders, log: LoggingAdapter) =
+    //todo limit max branch depth?
+    if (message.headers.nonEmpty && message.headers.head.hash == headers.head.parentHash) {
+      headers = message.headers.reverse ++ headers
+      processBlockHeaders(headers, context.sender(), log)
+    } else {
+      //we did not get previous blocks, there is no way to resolve, blacklist peer and continue download
+      resumeWithDifferentPeer()
+    }
+
+  private def handleDownload(message: BlockHeaders, log: LoggingAdapter) = if (message.headers.nonEmpty) {
+    headers = message.headers.sortBy(_.number)
+    processBlockHeaders(headers, context.sender(), log)
+  } else {
+    //no new headers to process, schedule to ask again in future, we are at the top of chain
+    broadcasting = true
+    scheduleResume
+  }
+
+  private def processBlockHeaders(headers: Seq[BlockHeader], peer: ActorRef, log: LoggingAdapter) = {
+    val parentByNumber = blockchain.getBlockHeaderByNumber(headers.head.number - 1)
+
+    parentByNumber match {
+      case Some(parent) =>
+        //we have same chain
+        if (parent.hash == headers.head.parentHash) {
+          peer ! SendMessage(GetBlockBodies(headers.take(blockBodiesPerRequest).map(_.hash)))
+        } else {
+          context.self ! ResolveBranch(context.sender())
+        }
+      case None =>
+        log.warning("got header that does not have parent")
+        resumeWithDifferentPeer()
+    }
+  }
+
+  private def handleBlockBodies(m: BlockBodies, log: LoggingAdapter) = {
+    if (m.bodies.nonEmpty) {
+      val result = headers.zip(m.bodies).map { case (h, b) => blockValidator(h, b) }
```

Comment:
in this version I am just zipping headers with bodies and then perform validation, if block body will not match validation will fail, also zip will automatically pick same number of headers as we got bodies
also it is correct state if we have more headers then bodies, it means that we should ask for remaining bodies in next request